### ISSUE TITLE
Model quantification problem

### ISSUE THREAD

**Issue by DurhackDa** [andre1araujo/YOLO-on-PYNQ-Z2#13]
Hello author, thank you very much for sharing this project. When I ran decent_q inspect --input_frozen_graph=model_data/yolo.pb during the quantization stage, I got exactly the same results as you, as shown, 
![image](https://github.com/user-attachments/assets/f025c997-f380-489b-8efc-b0682b7ea2bb)
but when I followed your steps and carefully changed the position of the dataset, When running the script, I keep encountering the following problem. Have you ever encountered this problem?
![image](https://github.com/user-attachments/assets/74ed02ee-946c-4b94-bd2f-15647937a254)

**Comment by andre1araujo** [comment]
Hello!
I have not encountered that specific problem as far as I remember. In any case, from the error message, I suspect you are willing to use the GPU of your machine to accelerate the process, right?
If that is the case, you probably modified the `quant.sh` script placing a '1' on the GPU setting. The problem now is that you are probably using the CPU version of DNNDK. If you want to use the GPU version, you have to do [this pat](https://andre-araujo.gitbook.io/yolo-on-pynq-z2/model-optimization-and-compilation#dnndk-setup) again but using the gpu wheel file.
What I mean is that here, we did this command: `pip install ./tensorflow-1.12.0-cp36-cp36m-linux_x86_64.whl`
This command uses the "cp" meaning CPU. If you navigate the folder there is a GPU version you can use.
The command you should use is this: `pip install tensorflow_gpu-1.12.0-cp36-cp36m-linux_x86_64.whl`

If you are indeed trying to use the GPU , give me that information because if not I have to search for other causes of the error.
Good luck! :)

**Comment by DurhackDa** [comment]
I did not try to use the GPU. Did I use other versions of tensorflow when converting and generating yolo.pb, and then used version 1.12 in the quantization stage, which caused the content of yolo.pb to be unable to be read, leading to the current problem?

**Comment by andre1araujo** [comment]
The .pb file should be in Tensorflow verison 1.12 or lower. If you followed the previous steps, that is granted. Are you using a different version of YOLO?
Can you show me your quant.sh script and the graph_input_fn.py script?

**Comment by DurhackDa** [comment]
It is the same as what you are using in the project, but when I convert the yolo model from Darknet to Keras and convert the Keras model to Tensorflow, I may be using other versions of tensorflow.

**Comment by andre1araujo** [comment]
Okay, that is probably it. Unfortunately one of the biggest limitations is the Tensorflow version DNNDK accepts. Try to make sure the version is lower than 1.12 and it should work without a problem. Newer Tensorflow versions have changes in some commands and syntax and that is probably why the error is referring to a padding issue as on a newer version that might be coded differently.
I hope this helps. Let me know of further developments!

**Comment by DurhackDa** [comment]
Hello, I have made some modifications, but now I have encountered a new problem. When I use the dataset to train the model, I get the following error message, which seems to be due to insufficient memory. I would like to ask, how much memory did you use when you were training? Are there any rigid requirements?
![image](https://github.com/user-attachments/assets/dbc4f339-0f81-4504-8649-7b687d1ca08f)

**Comment by andre1araujo** [comment]
Hi again!
My computer has about 12GB of memory. I was able to do it without any warnings. If the problem is indeed related to memory, you can just reduce the batch size for the quantization. That can be done by opening the calibration script called "graph_input_fn.py" and change the `calib_batch_size` to a smaller number. The default is 50, so you can experiment, for example with 30 or 20 and see if the error persists. 

![image](https://github.com/user-attachments/assets/036811aa-332e-4929-8d9e-d5bc3e1c8cb2)

If at batch size '1', the same error occurs it means the problem is related to something else.
I hope this helps you!
Good luck! :)

**Comment by DurhackDa** [comment]
I just tried it and it still doesn't work. Maybe my available memory is too low, only 3G size. I'll try to see if I can complete the training task on a different computer tomorrow, thank you very much for your reply!

**Comment by DurhackDa** [comment]
Hi, does this quantisation time have much to do with the number of images? I'm using 120 images to test the results first, is it possible that I don't need to wait as long as 12 hours.

**Comment by andre1araujo** [comment]
The number of images of the dataset directly relate to the time it takes for the quantization process. It needs images for calibration, so if the number is lower it will take less time but on the other hand it will make the model less accurate.
The quantization time is also related to the processing unit, your CPU. Of course if it is a more powerful CPU it will take less time. This process would be considerably faster with a GPU than a CPU.
I suggest you just leave it at night working on the quantization. Make sure it is on a well ventilated room to make sure nothing overheats on your computer. You might also raise your laptop a little bit to let more air get in the intake vents.

**Comment by DurhackDa** [comment]
My process is still terminated, can I use 20,000 photos and change calib_iter 100 to 20 and give up some accuracy.

**Comment by DurhackDa** [comment]
I'm really sorry, I still can't finish the training, do you have a GENERATED QUANTIZE FILE that has been trained that I can use and continue with the project later?

发送自 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: André Araújo ***@***.***>
发送时间: Tuesday, August 20, 2024 7:03:51 PM
收件人: andre1araujo/YOLO-on-PYNQ-Z2 ***@***.***>
抄送: CUI, DA (Student) ***@***.***>; Author ***@***.***>
主题: Re: [andre1araujo/YOLO-on-PYNQ-Z2] Model quantification problem (Issue #13)

[EXTERNAL EMAIL]

The number of images of the dataset directly relate to the time it takes for the quantization process. It needs images for calibration, so if the number is lower it will take less time but on the other hand it will make the model less accurate.
The quantization time is also related to the processing unit, your CPU. Of course if it is a more powerful CPU it will take less time. This process would be considerably faster with a GPU than a CPU.
I suggest you just leave it at night working on the quantization. Make sure it is on a well ventilated room to make sure nothing overheats on your computer. You might also raise your laptop a little bit to let more air get in the intake vents.

―

**Comment by andre1araujo** [comment]
Yes, you can use the 20000 photos for the quantization and reduce the calibration batch size to help speed up the process.
Yes, I have the quantization results for you if you want to skip that step. You can access the files on [this Google drive](https://drive.google.com/drive/folders/11gCSzDKbSrAvArS232TQK5Ga-qBcJR5m). 
I hope this helps and sorry for any inconvenience!

**Comment by DurhackDa** [comment]
Thank you very much for sharing. I will continue with the next steps and see what caused the failure later.
________________________________
发件人: André Araújo ***@***.***>
发送时间: 2024年8月20日 22:04
收件人: andre1araujo/YOLO-on-PYNQ-Z2 ***@***.***>
抄送: CUI, DA (Student) ***@***.***>; Author ***@***.***>
主题: Re: [andre1araujo/YOLO-on-PYNQ-Z2] Model quantification problem (Issue #13)

[EXTERNAL EMAIL]

Yes, you can use the 20000 photos for the quantization and reduce the calibration batch size to help speed up the process.
Yes, I have the quantization results for you if you want to skip that step. You can access the files on this Google drive<https://drive.google.com/drive/folders/11gCSzDKbSrAvArS232TQK5Ga-qBcJR5m>.
I hope this helps and sorry for any inconvenience!

―

**Comment by DurhackDa** [comment]
![image](https://github.com/user-attachments/assets/dc6f9c7c-4a30-419c-be6c-dcc5ea5a8617)
I would like to ask where the zynq7020 dnndk v3.1 mentioned in the picture is? Sorry I didn't find it.

**Comment by DurhackDa** [comment]
I found two small mistakes： 
(1) Copy the contents of the folder to the SD card here. 
![image](https://github.com/user-attachments/assets/4ae6030b-7c2e-4ff8-a33b-4307296f004b)
There will be no zynq7020_dnndk_v3.1 folder here, just open root directly 
![image](https://github.com/user-attachments/assets/9089d4c3-4348-4b22-ad90-7be7b1dbe5f4)
(2) The IP address here should be ping 192.168.2.99
![image](https://github.com/user-attachments/assets/2e7161a2-9005-4e6b-ac26-293ceec5d5a5)

**Comment by andre1araujo** [comment]
Hi!
I did not understand correctly the first mistake you are pointing out. My English is not perfect at all haha.
In any case, the zynq7020_dnndk_v3.1 folder is located [here](https://github.com/andre1araujo/YOLO-on-PYNQ-Z2/tree/main/DPU%20implementation/Petalinux).
The second part I noticed it was 192.168.2.99 instead of 192.168.2.9. I already corrected it but it was no big mistake since both addresses should work!
Thank you very much for pointing out mistakes!

**Comment by DurhackDa** [comment]
Hello, the first problem is that if there is * in this command, zynq7020_dnndk_v3.1 will not be created, and the file will not be found in the second command. I have a more major problem now. After I used deploy-model.pb to generate the elf file, there was only one result for image recognition in the end. I used NETRO to check the .pb file and found that it only had one solution, cov2d_59, and no 67 and 75, is this pb the result of complete training?
![image](https://github.com/user-attachments/assets/6e0e1341-2794-490a-9846-ff47c2328bee)

**Comment by andre1araujo** [comment]
Hello!
What do you mean with "only one result for image recognition"? It only detects one class of objects?
I might have provided the wrong file but I am sure that is not the case. The .pb file was the result of 9 hours of quantization using the COCO dataset images. 
Did you proceed with that file to the end and got any wrong results?

**Comment by DurhackDa** [comment]
Yes, it only detects the bicycle in the images

发送自 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: André Araújo ***@***.***>
发送时间: Wednesday, August 21, 2024 10:51:53 AM
收件人: andre1araujo/YOLO-on-PYNQ-Z2 ***@***.***>
抄送: CUI, DA (Student) ***@***.***>; Author ***@***.***>
主题: Re: [andre1araujo/YOLO-on-PYNQ-Z2] Model quantification problem (Issue #13)

[EXTERNAL EMAIL]

Hello!
What do you mean with "only one result for image recognition"? It only detects one class of objects?
I might have provided the wrong file but I am sure that is not the case. The .pb file was the result of 9 hours of quantization using the COCO dataset images.

―

**Comment by andre1araujo** [comment]
It should work better than that. If you test it on other images, it will detect more objects. It just seems to be lacking the precision it normally has. 
Try instead to use [these compilation files](https://drive.google.com/drive/u/2/folders/1vjVQ8fHe-IJwtoLmnupZuPJlAxdrQD3b) I generated and see if it works better.

**Comment by DurhackDa** [comment]
Hi, I would like to ask, in this label position in the figure, is this coordinate obtained under the original size of the validation set image or the size normalized to 416*416 size?
![image](https://github.com/user-attachments/assets/071bd89d-d726-48a0-bc5e-1a80a2109701)

**Comment by andre1araujo** [comment]
Hello!
Sorry for the delayed response.
**The coordinates are relative to the validation image with the original size** and not the resized image. The original image is resized to 416x416 pixels using the _'letterbox_image'_ function, sent to the YOLO neural network for the detections and then passes the _'correct_region_boxes'_ function to resize it back to the original size. Only then the _'deal'_ function is called, which registers the coordinates and objects.
I hope this helps!

**Comment by andre1araujo** [comment]
Hello!
I will be closing this issue for inactivity. If you desire to continue reporting issues make sure you open a new one!
I am sorry for the inconvenience and  thank you for understanding :)
